docs/site/todo-tutorial-geocoding-service.md

```diff
@@ -0,0 +1,260 @@
+consume e.g. from Controllers.
+
+In LoopBack, each service proxy is backed by a
+[DataSource](./todo-tutorial-datasource.md), this datasource leverages one of
+the service connectors to make outgoing requests and parse responses returned by
+the service.
+
+In our tutorial, we will leverage
+[US Census Geocoder API](https://geocoding.geo.census.gov/geocoder/) to convert
+textual US addresses into GPS coordinates, thus enabling client applications of
+our Todo API to display location-based reminders,
+
+{% include tip.html content="
+In a real project, you may want to use a geocoding service that covers more
+countries beyond USA and provides faster responses than US Census Geocoder API,
+for example IBM's [Weather Company Data](https://console.bluemix.net/catalog/services/weather-company-data)
+or [Google Maps Platform](https://developers.google.com/maps/documentation/geocoding).
+" %}
+
+### Configure the backing datasource
+
+Run `lb4 datasource` to define a new datasource connecting to Geocoder REST
+service.
+
+```
+$ lb4 datasource
+? Datasource name: geocoder
+? Select the connector for geocoder: REST services (supported by StrongLoop)
+? Base URL for the REST service:
+? Default options for the request:
+? An array of operation templates:
+? Use default CRUD mapping: No
+   create src/datasources/geocoder.datasource.json
+   create src/datasources/geocoder.datasource.ts
+ # npm will install dependencies now
+    update src/datasources/index.ts
+
+Datasource geocoder was created in src/datasources/
+```
+
+Edit the newly created datasource configuration to configure Geocoder API
+endpoints.
+
+#### /src/datasources/geocoder.datasource.json
+
+```json
+{
+  "connector": "rest",
+  "options": {
+    "headers": {
+      "accept": "application/json",
+      "content-type": "application/json"
+    }
+  },
+  "operations": [
+    {
+      "template": {
+        "method": "GET",
+        "url":
+          "https://geocoding.geo.census.gov/geocoder/locations/onelineaddress",
+        "query": {
+          "format": "{format=json}",
+          "benchmark": "Public_AR_Current",
+          "address": "{address}"
+        },
+        "responsePath": "$.result.addressMatches[*].coordinates"
+      },
+      "functions": {
+        "geocode": ["address"]
+      }
+    }
+  ]
+}
+```
+
+### Implement a service provider
+
+Create a new directory `src/services` and add the following two new files:
+
+#### src/geocoder.service.ts
+
+```ts
+import {getService, juggler} from '@loopback/service-proxy';
+import {inject, Provider} from '@loopback/core';
+import {GeocoderDataSource} from '../datasources/geocoder.datasource';
+
+export interface GeoPoint {
+  /**
+   * latitude
+   */
+  y: number;
+
+  /**
+   * longitude
+   */
+  x: number;
+}
+
+export interface GeocoderService {
+  geocode(address: string): Promise<GeoPoint[]>;
+}
+
+export class GeocoderServiceProvider implements Provider<GeocoderService> {
```

Comment:
I wonder why we don't use `@serviceProxy` - http://loopback.io/doc/en/lb4/Calling-other-APIs-and-web-services.html#declare-service-proxies-for-your-controller.

Reply:
When I started to look into testing services, I found `@serviceProxy` to make testing difficult, because tests cannot use the same code path for obtaining the service proxy instance as the production code uses.

See this integration test for an example:

https://github.com/strongloop/loopback-next/blob/9d82c9e67faba57a7d274f1231c2d11548dd7629/examples/todo/test/integration/services/geocoder.service.integration.ts#L37-L41

I am of the opinion that integration tests should not be using Application, Context and Dependency Injection, instead they should instantiate all classes directly from code. This makes refactorings easier, because all dependencies and couplings are explicitly expressed in code and IDE tooling can update them accordingly. (Compare that to acceptance tests where a lot of configuration is based on convention and string names.)

Reply:
If we did not need to tweak datasource settings to add caching proxy, the code would become even more simple:

```ts
const dataSource = new GeocoderDataSource(); 
service = new GeocoderServiceProvider(dataSource).value();
```

Originally, before support for booting datasources was added, this used to be a one-liner. I think we may be able to get there using default arguments in the current design too.

```ts
service = new GeocoderServiceProvider().value();
```

> At the moment, services are backed/configured by datasources. If we want to promote service as the 1st class artifact, we probably should evaluate the possibility of dividing datasource into two types:
>
> - datasource
> - service

I am pretty happy about the current design, to be honest.
-  A datasource is a long-living singleton-like object that manages state shared by all incoming HTTP requests - PersitedModel definitions, connection pool, etc.
- A `Repository` or a `Service` is a lightweight cheap-to-create facade exposing user-friendly API to controllers, leveraging datasources to do the heavy lifting.

In my mind, a Service (or a Service Proxy) is similar to a Repository.

@raymondfeng thoughts?

Reply:
I am going to land this pull request. If there are any changes identified in the outcome of this discussion, then I'll open a new pull request to address them.

Such changes are likely to affect the following places:
 - `examples/todo` implementation
 - "Calling other services" docs
 - "Testing your application"  docs